drivers/bluetooth/hci/hci_esp32.c

```diff
@@ -0,0 +1,313 @@
+	} else {
+		BT_ERR("No available ACL buffers!");
+		return NULL;
+	}
+
+	if (remaining != sys_le16_to_cpu(hdr.len)) {
+		BT_ERR("ACL payload length is not correct");
+		net_buf_unref(buf);
+		return NULL;
+	}
+
+	BT_DBG("len %u", remaining);
+	net_buf_add_mem(buf, data, remaining);
+
+	return buf;
+}
+
+static struct net_buf *bt_esp_iso_recv(uint8_t *data, size_t remaining)
+{
+	struct bt_hci_iso_hdr hdr;
+	struct net_buf *buf;
+
+	if (remaining < sizeof(hdr)) {
+		BT_ERR("Not enough data for ISO header");
+		return NULL;
+	}
+
+	buf = bt_buf_get_rx(BT_BUF_ISO_IN, K_NO_WAIT);
+	if (buf) {
+		memcpy((void *)&hdr, data, sizeof(hdr));
+		data += sizeof(hdr);
+		remaining -= sizeof(hdr);
+
+		net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	} else {
+		BT_ERR("No available ISO buffers!");
+		return NULL;
+	}
+
+	if (remaining != sys_le16_to_cpu(hdr.len)) {
+		BT_ERR("ISO payload length is not correct");
+		net_buf_unref(buf);
+		return NULL;
+	}
+
+	BT_DBG("len %zu", remaining);
+	net_buf_add_mem(buf, data, remaining);
+
+	return buf;
+}
+
+static int hci_esp_host_rcv_pkt(uint8_t *data, uint16_t len)
+{
+	uint8_t pkt_indicator;
+	struct net_buf *buf = NULL;
+	size_t remaining = len;
+
+	BT_HEXDUMP_DBG(data, len, "host packet data:");
+
+	pkt_indicator = *data++;
+	remaining -= sizeof(pkt_indicator);
+
+	switch (pkt_indicator) {
+	case HCI_EVT:
+		buf = bt_esp_evt_recv(data, remaining);
+		break;
+
+	case HCI_ACL:
+		buf = bt_esp_acl_recv(data, remaining);
+		break;
+
+	case HCI_SCO:
+		buf = bt_esp_iso_recv(data, remaining);
+		break;
+
+	default:
+		BT_ERR("Unknown HCI type %u", pkt_indicator);
+		return -1;
+	}
+
+	if (buf) {
+		BT_DBG("Calling bt_recv(%p)", buf);
+
+		bt_recv(buf);
+	}
+
+	return 0;
+}
+
+static void hci_esp_controller_rcv_pkt_ready(void)
+{
+	k_sem_give(&hci_send_sem);
+}
+
+static esp_vhci_host_callback_t vhci_host_cb = {
+	hci_esp_controller_rcv_pkt_ready,
+	hci_esp_host_rcv_pkt
+};
+
+static int bt_esp32_send(struct net_buf *buf)
+{
+	int err = 0;
+	uint8_t pkt_indicator;
+
+	BT_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
+
+	switch (bt_buf_get_type(buf)) {
+	case BT_BUF_ACL_OUT:
+		pkt_indicator = HCI_ACL;
+		break;
+	case BT_BUF_CMD:
+		pkt_indicator = HCI_CMD;
+		break;
+	case BT_BUF_ISO_OUT:
+		pkt_indicator = HCI_ISO;
+		break;
+	default:
+		BT_ERR("Unknown type %u", bt_buf_get_type(buf));
+		goto done;
+	}
+	net_buf_push_u8(buf, pkt_indicator);
+
+	BT_HEXDUMP_DBG(buf->data, buf->len, "Final HCI buffer:");
+
+	if (!esp_vhci_host_check_send_available()) {
+		BT_WARN("Controller not ready to receive packets");
+	}
+
+	if (k_sem_take(&hci_send_sem, HCI_BT_ESP32_TIMEOUT) == 0) {
+		esp_vhci_host_send_packet(buf->data, buf->len);
+	} else {
+		BT_ERR("Send packet timeout error");
+		err = -ETIMEDOUT;
+	}
+
+done:
+	net_buf_unref(buf);
+	k_sem_give(&hci_send_sem);
+
+	return err;
+}
+
+static int bt_esp32_ble_init(void)
+{
+	int ret;
+	esp_bt_controller_config_t bt_cfg = BT_CONTROLLER_INIT_CONFIG_DEFAULT();
```

Comment:
Might want to be careful about using `BT_` as a prefix in your external headers, the rest use the `esp_bt` prefix which seems somewhat safer. I'm assuming this is from `esp_bt.h`